lib/shoehorn/release.ex

```diff
@@ -15,26 +15,170 @@ defmodule Shoehorn.Release do
+
+        other_mode ->
+          other_mode
+      end
+
+    {app, new_mode}
+  end
+
+  defp add_release_apps(dep_graph, release_apps) do
+    Enum.each(release_apps, fn {app, _info} -> :digraph.add_vertex(dep_graph, app) end)
+
+    Enum.each(release_apps, fn {app, info} ->
+      Enum.each(info[:applications], &:digraph.add_edge(dep_graph, app, &1, :release))
+    end)
+
+    dep_graph
   end
 
-  def start_apps(applications) do
-    Enum.filter(applications, fn {name, _opts} ->
-      name in @permanent_applications
+  defp add_extra_dependencies(dep_graph, extra_deps) do
+    Enum.each(extra_deps, fn {app, deps} ->
+      Enum.each(deps, &checked_add_edge(dep_graph, app, &1))
     end)
-    |> Enum.map(&elem(&1, 0))
-    |> Enum.map(&{&1, :permanent})
+
+    dep_graph
   end
 
-  def load_apps(applications) do
-    Enum.reject(applications, fn {name, _opts} ->
-      name in @permanent_applications
+  defp checked_add_edge(graph, app, dep) do
+    case :digraph.add_edge(graph, app, dep, :extra) do
+      {:error, {:bad_vertex, v}} ->
+        raise RuntimeError, "Unknown application #{inspect(v)}"
+
+      {:error, {:bad_edge, [_, _]}} ->
+        # Edge already exists, so this is ok
+        :ok
+
+      {:error, {:bad_edge, _path}} ->
+        raise RuntimeError,
+              "Cycle detected when adding the #{inspect(dep)} dependencies to #{inspect(app)}"
+
+      _ ->
+        :ok
+    end
+  end
+
+  defp add_init_dependencies(dep_graph, init_apps) do
+    # Make every non-init_app depend on the init_app unless there's a cycle
+    all_apps = :digraph.vertices(dep_graph)
+    non_init_apps = all_apps -- init_apps
+
+    # Order deps in the init list
+    order_dependencies(dep_graph, Enum.reverse(init_apps))
+
+    # Try to make everything not in the init list depend on the init list
+    # (cycles and dupes are automatically ignored)
+    Enum.each(non_init_apps, fn non_init_app ->
+      Enum.each(init_apps, &:digraph.add_edge(dep_graph, non_init_app, &1, :init))
     end)
-    |> Enum.map(&elem(&1, 0))
-    |> Enum.map(&{&1, :none})
+
+    dep_graph
+  end
+
+  defp add_last_dependencies(dep_graph, last_apps) do
+    # Make every last_app depend on all non-last_apps unless there's a cycle
+    all_apps = :digraph.vertices(dep_graph)
+    non_last_apps = all_apps -- last_apps
+
+    Enum.each(last_apps, fn last_app ->
+      Enum.each(non_last_apps, &:digraph.add_edge(dep_graph, last_app, &1, :last))
+    end)
+
+    dep_graph
+  end
+
+  defp alphabetize_dependencies(dep_graph) do
+    # Add edges where possible to force dependencies to be sorted alphabetically
+    sorted_apps = :digraph.vertices(dep_graph) |> Enum.sort(:desc)
+
+    order_dependencies(dep_graph, sorted_apps)
+
+    dep_graph
+  end
+
+  # defp inspect_graph(dep_graph) do
+  #   Enum.each(:digraph.edges(dep_graph), fn e ->
+  #     {_, v1, v2, label} = :digraph.edge(dep_graph, e)
+  #     IO.puts("#{v1} -> #{v2} (#{label})")
+  #   end)
+  # end
```

Comment:
Is this needed?

Reply:
Not needed. I just was sad to remove it quickly since you can just inspect a digraph to see it.